### ISSUE TITLE
Mnist example trainer evaluator split

### ISSUE THREAD

**Issue by oteph** [pytorch/ignite#678]
The mnist example creates a trainer and a sperate evaluator.

```py    
trainer = create_supervised_trainer(model, optimizer, F.nll_loss, device=device)
evaluator = create_supervised_evaluator(model,
                                            metrics={'accuracy': Accuracy(),
                                                     'nll': Loss(F.nll_loss)},
                                            device=device)
```

The evaluator is the used to calucalte the loss over the whole batch:

```py 
    @trainer.on(Events.EPOCH_COMPLETED)
    def log_training_results(engine):
        pbar.refresh()
        evaluator.run(train_loader)
        metrics = evaluator.state.metrics
        avg_accuracy = metrics['accuracy']
        avg_nll = metrics['nll']
        tqdm.write(
            "Training Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}"
            .format(engine.state.epoch, avg_accuracy, avg_nll)
        )
```

In this configuration the training sets gets evaluated twice. This slows down training. Also if the training data is randomized in any way -`shuffle=True` - (or especially with random augmentations), the data the evaluator sees is different form the data that went through the trainer. This causes a missmatch between the losses written per iteration and the loss given as the result of the training.

If users use this example as a starting point this can cause all kinds of issues later on (as it did for me).

**Comment by vfdev-5** [comment]
> In this configuration the training sets gets evaluated twice.

@oteph, well, more precisely, it is evaluated once with fixed weights model. During the training model weights are updated and metrics computed during the training do not represent the latest model's performances. And this also causes the differences for 
> This causes a missmatch between the losses written per iteration and the loss given as the result of the training.


Concerning 
> Also if the training data is randomized in any way -shuffle=True - (or especially with random augmentations), the data the evaluator sees is different form the data that went through the trainer.

This is correct and in more serious cases we can think to provide 3 data loaders:
- random data augmented training data loader for training
- fixed training subset data loader for evaluation
- fixed validation data loader for evaluation

Two last loader can provide the insights on model's overfitting, for example.

**Comment by vfdev-5** [comment]
@oteph I close this issue as answered. Please feel free to reopen if needed
